splicemachine/spliceengine
# DB-10621 Batch DDL Changes in drop schema cascade

## Short Description
Improve drop schema cascade performance

## Long Description
Drop schema cascade was slow for two reasons:
1. DB-9553 introduced a savepoint in TabInfoImpl.deleteRows, so we would effectively create a child transaction for every single delete that is caused by schema cascade, and we ended up with a transaction nesting level as large as the number of items to drop. We can now remove that savepoint as SICompactionStateMutate prevents us from purging the antitombstone if there is a tombstone with the same timestamp. Removing that savepoint brought `drop schema foo cascade` where foo contains 1000 tables down to 150-200 seconds on standalone. It used to crash before that point.
2. Each time we drop an object, we issue an RPC call to other region servers to notify them of the metadata change. This is especially important for them to invalidate data dictionary cache. Instead, we now group DDLChange into independent messages that are sent to other region servers. This brought `drop schema foo cascade` where foo contains 1000 tables down to  20-30 seconds on standalone.

Example of dropping objects buckets by bucket:
```
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
x                 x                           Bucket1  x
x Bucket2 ┌──────┐x                 ┌──────┐           x
x         │Table │x                 │Table │           x
x         └──────┘x                 └──────┘◄────┐     x
x          ▲    ▲ x                  ▲           │     x
x          │ xxx│xx        xxxxxxxxxx│xxxxxxxxxxx│xxxxxxxxxx
xxxxxxxxxxx│xx  │          x         │           │         x
x  ┌───────┴┐x ┌┴────┐     x      ┌──┴────┐     ┌┴──────┐  x
x  │Trigger │x │View │     x      │Trigger│     │Trigger│  x
x  └────────┘x └─────┘     x      └───────┘     └───────┘  x
x            x  ▲          x   xxxxxxxxxxxxxxxxxxxxxxxxxxxxx
x            x  │          x   x
x            xxx│xxxxxxxxxxx   x
x              ┌┴────┐         x
x              │View │         x
x  Bucket0     └─────┘         x
x                              x
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

```

## How to test
```
create schema s;
create table s.foo0 (a int);
create table s.foo1 (a int);
create table s.foo2 (a int);
create table s.foo3 (a int);
create table s.foo4 (a int);
create table s.foo5 (a int);
create table s.foo6 (a int);
create table s.foo7 (a int);
create table s.foo8 (a int);
...
create table s.foo991 (a int);
create table s.foo992 (a int);
create table s.foo993 (a int);
create table s.foo994 (a int);
create table s.foo995 (a int);
create table s.foo996 (a int);
create table s.foo997 (a int);
create table s.foo998 (a int);
create table s.foo999 (a int);
drop schema s cascade;
```
should now return promptly.

## Follow-up work
DB-12040 tracks future work:
Currently, drop table still drops underlying dependencies automatically, so those dependencies are not included in the grouped DDLChange, so each dependency issues a separate RPC call to other region servers. `drop schema cascade` should explicitly drop all table dependencies before dropping the table
